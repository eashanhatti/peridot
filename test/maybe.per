let {
  'val Maybe
    : '[_ : OType] -> OType
    = '\A {
        signature {
          t : Bool,
          d :
            bool_elim t {
              A;
              signature {};
            }
        }
      };
  'val map
    : '[A : OType] -> '[_ : '[_ : A] -> A] -> '[_ : <Maybe A>] -> <Maybe A>
    = '\A f m {
        bool_elim (#t m)
        {
          structure {
            t = tt,
            d = <f (#d m)>
          };
          structure {
            t = ff,
            d = structure {}
          };
        }
      };
} in {
  OType
}
