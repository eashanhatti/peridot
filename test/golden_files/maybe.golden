((fromList
    [(UVGlobal 1000,
      Just
        (Neutral
           _
           (TwoElim
              (Neutral
                 _
                 (RecElim
                    (LocalVar Level {unLevel = 2})
                    (Field {unField = "tag"})))
              (LocalVar Level {unLevel = 0})
              (RecType fromList [])))),
     (UVGlobal 1001,
      Just (LocalVar Level {unLevel = 0}))],
  fromList [],
  fromList []),
 Declare
   Obj
   (Rigid (NameObjIntro (Id {unId = 0})))
   (ObjFunType
      Explicit
      (Rigid (TypeType Obj))
      (Rigid (TypeType Obj)))
   (Define
      (Rigid (NameObjIntro (Id {unId = 0})))
      (ObjFunIntro
         (RecType
            (fromList
               [(Field {unField = "tag"}, Rigid TwoType),
                (Field {unField = "data"},
                 TwoElim
                   (LocalVar (Index {unIndex = 0}))
                   (LocalVar (Index {unIndex = 1}))
                   (RecType (fromList [])))])))
      (Declare
         Obj
         (Rigid (NameObjIntro (Id {unId = 1})))
         (ObjFunType
            Unification
            (Rigid (TypeType Obj))
            (ObjFunElim
               (GlobalVar (Id {unId = 0}))
               (LocalVar (Index {unIndex = 0}))))
         (Declare
            Obj
            (Rigid (NameObjIntro (Id {unId = 2})))
            (ObjFunType
               Unification
               (Rigid (TypeType Obj))
               (ObjFunType
                  Explicit
                  (LocalVar (Index {unIndex = 0}))
                  (ObjFunElim
                     (GlobalVar (Id {unId = 0}))
                     (LocalVar (Index {unIndex = 1})))))
            (Declare
               Obj
               (Rigid (NameObjIntro (Id {unId = 3})))
               (ObjFunType
                  Explicit
                  (Rigid (TypeType Obj))
                  (ObjFunType
                     Explicit
                     (ObjFunType
                        Explicit
                        (LocalVar (Index {unIndex = 0}))
                        (LocalVar (Index {unIndex = 1})))
                     (ObjFunType
                        Explicit
                        (ObjFunElim
                           (GlobalVar (Id {unId = 0}))
                           (LocalVar (Index {unIndex = 1})))
                        (ObjFunElim
                           (GlobalVar (Id {unId = 0}))
                           (LocalVar (Index {unIndex = 2}))))))
               (Define
                  (Rigid (NameObjIntro (Id {unId = 1})))
                  (ObjFunIntro
                     (RecIntro
                        (fromList
                           [(Field {unField = "tag"}, Rigid TwoIntro1),
                            (Field {unField = "data"},
                             RecIntro (fromList []))])))
                  (Define
                     (Rigid (NameObjIntro (Id {unId = 2})))
                     (ObjFunIntro
                        (ObjFunIntro
                           (RecIntro
                              (fromList
                                 [(Field {unField = "tag"}, Rigid TwoIntro0),
                                  (Field {unField = "data"},
                                   LocalVar (Index {unIndex = 1}))]))))
                     (Define
                        (Rigid (NameObjIntro (Id {unId = 3})))
                        (ObjFunIntro
                           (ObjFunIntro
                              (ObjFunIntro
                                 (TwoElim
                                    (RecElim
                                       (LocalVar (Index {unIndex = 0}))
                                       (Field {unField = "tag"}))
                                    (ObjFunElim
                                       (ObjFunElim
                                          (GlobalVar (Id {unId = 2}))
                                          (UniVar (UVGlobal 1000)))
                                       (RecElim
                                          (LocalVar (Index {unIndex = 0}))
                                          (Field {unField = "data"})))
                                    (ObjFunElim
                                       (GlobalVar (Id {unId = 1}))
                                       (UniVar (UVGlobal 1001)))))))
                        (Rigid (TypeType Obj))))))))))
